File: Java/nextPermutation.java
public class nextPermutation {
    public static void solution(int[] nums) {
        int n = nums.length, i, j;

        // Step 1: Find first decreasing element from right
        for (i = n - 2; i >= 0; i--) {
            if (nums[i] < nums[i + 1])
                break;
        }

        // Step 2: Find element just larger than nums[i] and swap
        if (i >= 0) {
            for (j = n - 1; j > i; j--) {
                if (nums[j] > nums[i]) {
                    swap(nums, i, j);
                    break; // Swap done, exit loop
                }
            }
        }

        // Step 3: Reverse the elements after index 'i' to get next permutation
        reverse(nums, i + 1, n - 1);
    }

    public static void swap(int[] nums, int i, int j) {
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }

    public static void reverse(int nums[], int start, int end) {
        while (start < end) {
            swap(nums, start, end);
            start++;
            end--;
        }
    }

    public static void main(String[] args) {
        int[] nums = {1, 2, 3}; // Example input
        solution(nums);
        
        System.out.print("Next Permutation: ");
        for (int num : nums) {
            System.out.print(num + " ");
        }
        System.out.println();
    }
}

File: Python/nextPermutation.py
def nextPermutation(nums):
    n = len(nums)
    
    # Step 1: Find the first decreasing element from the end
    i = n - 2
    while i >= 0 and nums[i] >= nums[i + 1]:
        i -= 1
    
    # Step 2: If the entire array is not in descending order
    if i >= 0:
        # Find the element just larger than nums[i] and swap
        j = n - 1
        while nums[j] <= nums[i]:
            j -= 1
        nums[i], nums[j] = nums[j], nums[i]  # Swap
    
    # Step 3: Reverse the elements after index 'i'
    nums[i + 1:] = reversed(nums[i + 1:])
    
# Example usage
nums = [1, 2, 3]
nextPermutation(nums)
print("Next permutation:", nums)
